src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -991,45 +999,86 @@
+                                            :provider provider
+                                            :hash (sutils/munge-hash-for-storage package-hash)})
+                                         packages-to-create))]
+    (merge existing-hashes-map
+           (zipmap (map facts/package-tuple-hash packages-to-create)
+                   (map :id results)))))
+
 (s/defn update-packages
   "Compares `inventory` to the stored package inventory for
   `certname`. Differences will result in updates to the database"
   [certname
    inventory :- [facts/package-tuple]]
-  (let [{certname-id :id package_hash :package_hash} (find-certname-id-and-hash certname)
-        new-package-hash (shash/package-similarity-hash inventory)]
-
-    (when-not (= new-package-hash package_hash)
-      (let [old-packages (create-package-map certname-id)
-            [to-add to-remove _] (clojure.data/diff (set inventory) (set (keys old-packages)))
-            to-remove-ids (map old-packages to-remove)]
+  (let [{certname-id :id package-hash :package_hash} (find-certname-id-and-hash certname)
+        hashed-package-tuples (map shash/package-identity-hash inventory)
+        new-package-hash (shash/package-similarity-hash hashed-package-tuples)]
+
+    (when-not (= new-package-hash package-hash)
+      (let [just-hashes (map facts/package-tuple-hash hashed-package-tuples)
+            existing-package-hashes (find-package-hashes just-hashes)
+            full-hashes-map (insert-missing-packages existing-package-hashes hashed-package-tuples)
+            new-package-id-set (set (vals full-hashes-map)) ;(set (map full-hashes-map just-hashes))
+            [new-package-ids old-package-ids _] (clojure.data/diff new-package-id-set
+                                                                   (package-id-set-for-certname certname-id))]
         (jdbc/update! :certnames
                       {:package_hash (sutils/munge-hash-for-storage new-package-hash)}
                       ["id=?" certname-id])
-        (when (seq to-add)
-          (jdbc/insert-multi! :package_inventory
-                              ["certname_id" "name" "version" "provider"]
-                              (map #(cons certname-id %) to-add)))
-        (when (seq to-remove)
-          (jdbc/delete! :package_inventory
-                        ["id = ANY(?)"
-                         (sutils/array-to-param "bigint" Long to-remove-ids)]))))))
 
+        (when (seq new-package-ids)
+          (jdbc/insert-multi! :certname_packages
+                              ["certname_id" "package_id"]
+                              (map #(vector certname-id %) new-package-ids)))
+
+        (when (seq old-package-ids)
+          (jdbc/delete! :certname_packages
+                        ["certname_id = ? and package_id = ANY(?)"
+                         certname-id
+                         (sutils/array-to-param "bigint" Long (map long old-package-ids))]))))))
```

Comment:
Actually, I think this is a fairly good idea. It makes sure that we're definitely dealing with primitive longs. We may be implicitly boxing them in array-to-param, unsure though.